.github/workflows/tests.yml

```diff
@@ -6,46 +6,109 @@ on:
+        pip install -U wheel setuptools build
+
+    - name: Install requirements
+      run: pip install -r requirements.txt
+
+    - name: Lint
+      uses: suo/flake8-github-action@releases/v1
+      with:
+        checkName: 'flake8'
+      env:
+        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+
+    - name: Build
+      run: python -m build -o dist/
 
+    - uses: actions/upload-artifact@v3
+      with:
+        name: dist
+        path: dist
+
+  test:
+    runs-on: ubuntu-latest
+    needs: [build]
+    strategy:
+      matrix:
+        # todo: extract from source
+        python-version: [ 3.6, 3.7, 3.8, 3.9, '3.10' ]
+        install-from: ["dist/*.whl"]
     steps:
-      - name: Checkout to master
-        uses: actions/checkout@v2
-
-      - name: Setup python
-        uses: actions/setup-python@v2
-        with:
-          python-version: '3.x'
-          architecture: 'x64'
-
-      - name: Install
-        run: |
-          python -m pip install -e .
-          python -c "import pkg_resources; print(pkg_resources.get_distribution('crc-bonfire').version)"
-          which bonfire
-          bonfire --help
-          bonfire config write-default
-
-      - name: Install test reqs
-        run: pip install -r requirements.txt
-
-      - name: Lint
-        uses: suo/flake8-github-action@releases/v1
-        with:
-          checkName: 'flake8'
-        env:
-          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
-
-      - name: Unit tests
-        run: |
-          pytest -sv .
-
-      - name: Package build
-        run: |
-          python -m pip install --upgrade pip setuptools wheel build twine
-          python -m build
-          ls -l dist
-          python -m twine check dist/*
+    - uses: actions/checkout@v3
+    - name: Set up Python ${{ matrix.python-version }}
+      uses: actions/setup-python@v4
+      with:
+        python-version: ${{ matrix.python-version }}
+    - uses: actions/cache@v3
+      with:
+        path: ${{ env.pythonLocation }}
+        key: ${{ env.pythonLocation }}
+    - name: Install dependencies
+      run: |
+        python -m pip install --upgrade pip
+        pip install -U setuptools setuptools_scm pytest
+    - uses: actions/download-artifact@v3
+      with:
+        name: dist
+        path: dist
+    - name: install ${{ matrix.install-from }}
+      run: pip install ${{ matrix.install-from }}
+    - name: Install test requirements
+#      run: pip install -r requirements.txt
+#      TODO: refactor project's test requirements
+      run: pip install pytest-mock
+    - name: test bonfire version and create default config file
+      run: |
+        python -c "import pkg_resources; print(pkg_resources.get_distribution('crc-bonfire').version)"
+        which bonfire
+        bonfire --help
+        bonfire config write-default
+    - name: pytest
+      run: pytest -sv
 
+  build_check:
+    runs-on: ubuntu-latest
+    needs: [build]
+    steps:
+    - uses: actions/setup-python@v2
```

Comment:
As someone who isn't as familiar with Github actions, I was wondering if it would be more beneficial to use `actions/setup-python@v4` instead? This may not be required for what we are needing it for, but I figured I'd ask. This way we are also keeping things updated. Looks like it is being used in the `test` job.

Reply:
yep ! you're absolutely correct ! I saw an alert from the actions view that mentioned the Node version was going to make v2 to stop working

I've updated all the old references. Well, except the "flake8" action - I think I'm gonna refactor to not to use that at all - but on a follow-up PR

Thanks for mentioning it !